File: solutions/easy/704. Binary Search/binary_search.cpp
#include <vector>
#include <iostream> // Include this for testing purposes (optional)

using namespace std; // Add this line to avoid writing 'std::' before 'vector'

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int left = 0;
        int right = nums.size() - 1;
        int mid;
        while (left <= right) {
            mid = (right + left) / 2;
            if (target < nums[mid]) {
                right = mid - 1;
            } else if (target > nums[mid]) {
                left = mid + 1;
            } else {
                return mid;
            }
        }
        return -1;
    }
};

int main()
{
    vector<int> nums = {-1,0,3,5,9,12};
    int target = 9;
    Solution s;
    int result = s.search(nums, target);
    cout << "Index of target: " << result << endl;
    return 0;
}

File: solutions/easy/704. Binary Search/binary_search.py
from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1
        while left <= right:
            mid = (right + left) // 2
            if target < nums[mid]:
                right = mid - 1
            elif target > nums[mid]:
                left = mid + 1
            else:
                return mid
        return -1


def call(nums, target):
    s = Solution()
    return s.search(nums, target)


def test():
    assert call([-1, 0, 3, 5, 9, 12], 9) == 4
    assert call([-1, 0, 3, 5, 9, 12], 2) == -1
    assert call([], 2) == -1
    assert call([1], 1) == 0
